### PR TITLE
rpc: removed dependency on deprecated (*grpc.ClientConn).State()

### LINKED ISSUE #7430: client: remove dependency on (*grpc.ClientConn).State()
grpc-go has removed the `(*grpc.ClientConn).State()` function which we use for detection when a connection is down/ready to be used. We should probably remove all uses of this function so we no longer have to depend on an old version of grpc-go.

Related to grpc/grpc-go#690.

https://github.com/cockroachdb/cockroach/blob/master/client/rpc_sender.go#L55:L63

We removed the other use in #7399.

### PR DESCRIPTION
**Pull request by d4l3k**
Fixes #7430.

Tests fail with the updated dependencies.

```
E160720 16:46:41.551691 server.go:490  http2: server: error reading preface from client 127.0.0.1:49478: bogus greeting "GET /_status/metrics//1 "
I160720 16:46:41.551791 server/status_test.go:174  could not GET https://127.0.0.1:49466/_status/metrics//1 - Get https://127.0.0.1:49466/_status/metrics//1: EOF
--- FAIL: TestMetricsEndpoint (1.71s)
    status_test.go:194: There was an error retrieving https://127.0.0.1:49466/_status/metrics//1
```

Doesn't seem to be an issue when actually running a node.

---

This change is [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/7949)

### CODE REVIEW

**Comment by d4l3k** [comment]
Fixed the issue. A grpc update started set NextProtos on our global tls.Config object instead of making a copy with it.

https://github.com/grpc/grpc-go/commit/78e558bbc17465d318acd499a9610e120a837860

**Comment by tamird** [comment]
I don't think you can copy it like that; vet is going to complain. That's probably why upstream made the change begin with.

**Comment by d4l3k** [comment]
I thought so too but it isn't complaining for some reason

**Comment by tamird** [comment]
Reviewed 9 of 10 files at r3.
Review status: 0 of 17 files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[rpc/context.go, line 73 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7949#-KNEntx55CXoSNqOSw94:-KNEntx55CXoSNqOSw95:1172823038) ([raw file](https://github.com/cockroachdb/cockroach/blob/e470099f42599294e67255075cf0e490483797e6/rpc/context.go#L73)):*
> ```Go
> 
> const (
> 	// Unknown indicates that the health status of the remote server isn't known.
> ```

nit: you can have a single comment for this entire const block when the names are descriptive. In this case, these comments aren't really useful.

---

*[rpc/context.go, line 240 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7949#-KNEoOIaxW7mn4Pj73tZ:-KNEoOIaxW7mn4Pj73t_:1431421637) ([raw file](https://github.com/cockroachdb/cockroach/blob/e470099f42599294e67255075cf0e490483797e6/rpc/context.go#L240)):*
> ```Go
> // ConnState returns whether the most recent heartbeat succeeded or not. This
> // should not be used as a definite status of a nodes health and just used to
> // prioritized healthy nodes over unhealthy ones.
> ```

s/prioritized/prioritize/

---

*[rpc/context.go, line 248 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7949#-KNEoTuFCCoULuxRl_SJ:-KNEoTuFCCoULuxRl_SK:-164158216) ([raw file](https://github.com/cockroachdb/cockroach/blob/e470099f42599294e67255075cf0e490483797e6/rpc/context.go#L248)):*
> ```Go
> }
> 
> // ConnStateChange returns a channel that will trigger when the connection
> ```

"trigger"?

---

*[rpc/context.go, line 250 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7949#-KNEpb5d7_byDEzVKzfH:-KNEpb5eDTcCjXOKDIJ_:-1031015926) ([raw file](https://github.com/cockroachdb/cockroach/blob/e470099f42599294e67255075cf0e490483797e6/rpc/context.go#L250)):*
> ```Go
> // ConnStateChange returns a channel that will trigger when the connection
> // health state changes.
> func (ctx *Context) ConnStateChange(remoteAddr string) <-chan ConnectivityState {
> ```

instead of hanging these method on the context, can you make the dial method return a thin wrapper around the grpc.ClientConn that implements these methods? I don't see any reason to go back to the context every time.

I'd use that technique to just provide exactly the API that grpc gave us and continue to use that.

---

*[rpc/context.go, line 258 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7949#-KNEp51R3uNe2KsL71_I:-KNEp51R3uNe2KsL71_J:-1788793848) ([raw file](https://github.com/cockroachdb/cockroach/blob/e470099f42599294e67255075cf0e490483797e6/rpc/context.go#L258)):*
> ```Go
> 
> // WaitForConnected waits until the connection is no longer in the unknown
> // state. If the state is unhealthy, or times out it returns an error.
> ```

"error. state."

---

*[rpc/context.go, line 260 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7949#-KNEon1OK4dPOwdy4S-C:-KNEon1OK4dPOwdy4S-D:691401728) ([raw file](https://github.com/cockroachdb/cockroach/blob/e470099f42599294e67255075cf0e490483797e6/rpc/context.go#L260)):*
> ```Go
> // state. If the state is unhealthy, or times out it returns an error.
> // state.
> func (ctx *Context) WaitForConnected(remoteAddr string, timeout time.Duration) error {
> ```

seems like you could generalize this to exactly what grpc's WaitForStateChange provided.

---

*[rpc/context.go, line 263 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7949#-KNEoeIiNxBp379sk9YU:-KNEoeIiNxBp379sk9YV:1036739643) ([raw file](https://github.com/cockroachdb/cockroach/blob/e470099f42599294e67255075cf0e490483797e6/rpc/context.go#L263)):*
> ```Go
> 	var timerC <-chan time.Time
> 	if timeout > 0 {
> 		timerC = time.NewTimer(timeout).C
> ```

why not time.After? call it timeoutChan if you make the change.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7949)*

File: internal/client/rpc_sender.go [hunk 1]
```diff
@@ -24,47 +24,43 @@ import (
...
-
-	c := s.conn
-	for state, err := c.State(); state != grpc.Ready; state, err = c.WaitForStateChange(ctxWithTimeout, state) {
-		if err != nil {
-			return nil, roachpb.NewErrorf("roachpb.Batch RPC failed: %s", err)
-		}
-		if state == grpc.Shutdown {
-			return nil, roachpb.NewErrorf("roachpb.Batch RPC failed as client connection was closed")
-		}
+	if err := s.rpcContext.WaitForConnected(s.remoteAddr, healthyTimeout); err != nil {
```
Comment (tschottdorf) [inline internal/client/rpc_sender.go]: shouldn't `WaitForConnected` take the `context.Context` and fail early when it expires?

File: kv/transport.go [hunk 2]
```diff
@@ -178,16 +177,9 @@ func (gt *grpcTransport) SendNext(done chan BatchCall) {
...
-		for state, err := c.State(); state != grpc.Ready; state, err = c.WaitForStateChange(ctx, state) {
-			if err != nil {
-				done <- BatchCall{Err: err}
-				return
-			}
-			if state == grpc.Shutdown {
-				done <- BatchCall{Err: fmt.Errorf("rpc to %s failed as client connection was closed", addr)}
-				return
-			}
+		if err := gt.rpcContext.WaitForConnected(addr, gt.opts.Timeout); err != nil {
```
Comment (tschottdorf) [inline kv/transport.go]: Here, too.

**Comment by bdarnell** [comment]
Reviewed 4 of 5 files at r1, 1 of 10 files at r3, 9 of 9 files at r4, 1 of 1 files at r5, 2 of 2 files at r6.
Review status: all files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[security/tls.go, line 122 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7949#-KNTITENDpETAF78Wpas:-KNTITEOrzmZdypMcFxV:1166079319) ([raw file](https://github.com/cockroachdb/cockroach/blob/2dde8c3c84ad25e7018f8a0762825b7c6b781512/security/tls.go#L122)):*
> ```Go
> 		return cfg
> 	}
> 	newConfig := *cfg
> ```

Have you checked this with go 1.7 vet? It's gotten better about detecting copied locks. Whether vet complains or not, it's unsafe to use this unless you can be sure that no one has started using a non-copied version of this `tls.Config`. 

The Go standard library ran into this and solved it for themselves with a non-public method (grumble, grumble): https://github.com/golang/go/commit/d931716cde778a3a4c9ab14410f791e9e8b72785 (two subtly different non-public methods, in fact)

Other projects appear to have copied the same code (which will need to be updated for go 1.7 as two new fields have been added to the config struct). We could do the same but this really needs to be part of the standard library (See https://github.com/golang/go/issues/15771).  Or maybe we could pass around `func() *tls.Config` wrappers instead of pointers to shared `tls.Config` objects (eww, but the alternative is a brittle manual field-by-field copy).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7949)*

**Comment by d4l3k** [comment]
Review status: all files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[internal/client/rpc_sender.go, line 57 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7949#-KN9614MmjG7KVHQMdeh-r6-57:-KNYlLWdk1afxit8Ed01:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/2dde8c3c84ad25e7018f8a0762825b7c6b781512/internal/client/rpc_sender.go#L57)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> shouldn't `WaitForConnected` take the `context.Context` and fail early when it expires?
</details>
Done.

---

*[kv/transport.go, line 180 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7949#-KN9614MmjG7KVHQMdej-r6-180:-KNYZlsjcG3fgz6VBC7V:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/2dde8c3c84ad25e7018f8a0762825b7c6b781512/kv/transport.go#L180)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> Here, too.
</details>
Done.

---

*[rpc/context.go, line 73 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7949#-KNEntx55CXoSNqOSw94:-KNYZpGCKiqi5PY7H7nW:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/e470099f42599294e67255075cf0e490483797e6/rpc/context.go#L73)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> nit: you can have a single comment for this entire const block when the names are descriptive. In this case, these comments aren't really useful.
</details>
Done.

---

*[rpc/context.go, line 240 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7949#-KNEoOIaxW7mn4Pj73tZ:-KNY7MoG8jV_Y3bx4J7m:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/e470099f42599294e67255075cf0e490483797e6/rpc/context.go#L240)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> s/prioritized/prioritize/
</details>
Done.

---

*[rpc/context.go, line 248 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7949#-KNEoTuFCCoULuxRl_SJ:-KNYlHuEDKUwdZLu7fTt:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/e470099f42599294e67255075cf0e490483797e6/rpc/context.go#L248)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> "trigger"?
</details>
Done.

---

*[rpc/context.go, line 250 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7949#-KNEpb5d7_byDEzVKzfH:-KNYE3YiPexeFu5Mxfah:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/e470099f42599294e67255075cf0e490483797e6/rpc/context.go#L250)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> instead of hanging these method on the context, can you make the dial method return a thin wrapper around the grpc.ClientConn that implements these methods? I don't see any reason to go back to the context every time.
> 
> I'd use that technique to just provide exactly the API that grpc gave us and continue to use that.
</details>
Done.

---

*[rpc/context.go, line 258 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7949#-KNEp51R3uNe2KsL71_I:-KNYlK3UdEahliU1OhPO:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/e470099f42599294e67255075cf0e490483797e6/rpc/context.go#L258)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> "error. state."
</details>
Done.

---

*[rpc/context.go, line 260 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7949#-KNEon1OK4dPOwdy4S-C:-KNYE6uy5xdM7Uynbd6G:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/e470099f42599294e67255075cf0e490483797e6/rpc/context.go#L260)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> seems like you could generalize this to exactly what grpc's WaitForStateChange provided.
</details>
Done.

---

*[rpc/context.go, line 263 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7949#-KNEoeIiNxBp379sk9YU:-KNYE7MwCywhkj0PaxmN:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/e470099f42599294e67255075cf0e490483797e6/rpc/context.go#L263)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> why not time.After? call it timeoutChan if you make the change.
</details>
Done.

---

*[security/tls.go, line 122 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7949#-KNTITENDpETAF78Wpas:-KNXm383FlYFNMQuzkkH:2076887180) ([raw file](https://github.com/cockroachdb/cockroach/blob/2dde8c3c84ad25e7018f8a0762825b7c6b781512/security/tls.go#L122)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Have you checked this with go 1.7 vet? It's gotten better about detecting copied locks. Whether vet complains or not, it's unsafe to use this unless you can be sure that no one has started using a non-copied version of this `tls.Config`. 
> 
> The Go standard library ran into this and solved it for themselves with a non-public method (grumble, grumble): https://github.com/golang/go/commit/d931716cde778a3a4c9ab14410f791e9e8b72785 (two subtly different non-public methods, in fact)
> 
> Other projects appear to have copied the same code (which will need to be updated for go 1.7 as two new fields have been added to the config struct). We could do the same but this really needs to be part of the standard library (See https://github.com/golang/go/issues/15771).  Or maybe we could pass around `func() *tls.Config` wrappers instead of pointers to shared `tls.Config` objects (eww, but the alternative is a brittle manual field-by-field copy).
</details>
Updated to be a copy of (*tls.Config).clone() with a TODO about updating for 1.8. I can't imagine tls.Config fields changing between now and the 1.8 release.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7949)*

**Comment by bdarnell** [comment]
Review status: 0 of 30 files reviewed at latest revision, 10 unresolved discussions, some commit checks pending.

---

*[security/tls.go, line 122 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7949#-KNTITENDpETAF78Wpas:-KNYlsGlOCrzfHEYYd-X:821459457) ([raw file](https://github.com/cockroachdb/cockroach/blob/2dde8c3c84ad25e7018f8a0762825b7c6b781512/security/tls.go#L122)):*
<details><summary><i>Previously, d4l3k (Tristan Rice) wrote…</i></summary>
> Updated to be a copy of (*tls.Config).clone() with a TODO about updating for 1.8. I can't imagine tls.Config fields changing between now and the 1.8 release.
</details>
I can imagine it: https://github.com/golang/go/issues/16363 is proposing to add a new field and is tagged for the go1.8 milestone. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7949)*

**Comment by d4l3k** [comment]
*[security/tls.go, line 122 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7949#-KNTITENDpETAF78Wpas:-KNYnD9cLqweEe3F0sCY:-673505987) ([raw file](https://github.com/cockroachdb/cockroach/blob/2dde8c3c84ad25e7018f8a0762825b7c6b781512/security/tls.go#L122)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> I can imagine it: https://github.com/golang/go/issues/16363 is proposing to add a new field and is tagged for the go1.8 milestone. 
</details>
Sorry for the poor wording. `(*tls.Config).Clone()`is also slated for 1.8. When that gets added, we'll remove this function.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7949)*

**Comment by tamird** [comment]
Can you rebase this into two commits: 1bfe1db in one and everything else in another?

---

Reviewed 18 of 30 files at r7, 9 of 9 files at r8, 1 of 1 files at r9, 2 of 2 files at r10.
Review status: all files reviewed at latest revision, 10 unresolved discussions, some commit checks pending.

---

*[base/context.go, line 201 \[r10\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7949#-KNYmISE6IAOoHEI_aa_:-KNYmISE6IAOoHEI_aaa:-1457187179) ([raw file](https://github.com/cockroachdb/cockroach/blob/664ea88b6b5b8a6ec342dd4edf9549b0f8eb310e/base/context.go#L201)):*
> ```Go
> 	})
> 
> 	return security.CopyTLSConfig(ctx.clientTLSConfig.tlsConfig), ctx.clientTLSConfig.err
> ```

can you add comments explaining why?

---

*[internal/client/rpc_sender.go, line 33 \[r10\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7949#-KNYmp3TQp-oR2UI_4Fl:-KNYmp3TQp-oR2UI_4Fm:-1965831521) ([raw file](https://github.com/cockroachdb/cockroach/blob/664ea88b6b5b8a6ec342dd4edf9549b0f8eb310e/internal/client/rpc_sender.go#L33)):*
> ```Go
> 
> type sender struct {
> 	remoteAddr string
> ```

are these new fields still used? looks like no.

---

*[internal/client/rpc_sender.go, line 64 \[r10\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7949#-KNYmykEEyvtNc-bIVwI:-KNYmykEEyvtNc-bIVwJ:1478120799) ([raw file](https://github.com/cockroachdb/cockroach/blob/664ea88b6b5b8a6ec342dd4edf9549b0f8eb310e/internal/client/rpc_sender.go#L64)):*
> ```Go
> 	for state := c.State(); state != grpc.Ready; state, err = c.WaitForStateChange(ctxWithTimeout, state) {
> 		if err != nil {
> 			return nil, roachpb.NewError(errors.Wrap(err, "roachpb.Batch RPC failed"))
> ```

while you're here, give this a better message (WaitForStateChange failed)

---

*[kv/send_test.go, line 137 \[r10\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7949#-KNYnFKVaP4UPEiX2Hmo:-KNYnFKVaP4UPEiX2Hmp:2146093970) ([raw file](https://github.com/cockroachdb/cockroach/blob/664ea88b6b5b8a6ec342dd4edf9549b0f8eb310e/kv/send_test.go#L137)):*
> ```Go
> 	}
> 
> 	waitForConnState := func(desiredState grpc.ConnectivityState) {
> ```

why not use the method provided by `conn`?

---

*[rpc/client.go, line 35 \[r10\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7949#-KNYoPpr8ppv6T4CTpm9:-KNYoPpr8ppv6T4CTpmA:173233410) ([raw file](https://github.com/cockroachdb/cockroach/blob/664ea88b6b5b8a6ec342dd4edf9549b0f8eb310e/rpc/client.go#L35)):*
> ```Go
> }
> 
> // State returns whether the most recent heartbeat succeeded or not. This should
> ```

this comment implies a boolean return value, which is not the case.

---

*[rpc/client.go, line 47 \[r10\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7949#-KNYoZQNahBu1Cm8MEGi:-KNYoZQNahBu1Cm8MEGj:1246860652) ([raw file](https://github.com/cockroachdb/cockroach/blob/664ea88b6b5b8a6ec342dd4edf9549b0f8eb310e/rpc/client.go#L47)):*
> ```Go
> 		cc.state = grpc.Ready
> 	} else {
> 		cc.state = grpc.Shutdown
> ```

why shutdown?

---

*[rpc/context.go, line 69 \[r10\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7949#-KNYnV--AjA59a96-S90:-KNYnV--AjA59a96-S91:-496444635) ([raw file](https://github.com/cockroachdb/cockroach/blob/664ea88b6b5b8a6ec342dd4edf9549b0f8eb310e/rpc/context.go#L69)):*
> ```Go
> 
> type connMeta struct {
> 	conn *ClientConn
> ```

why is this a pointer? in fact, why does connMeta still exist at all?

---

*[security/tls.go, line 122 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7949#-KNTITENDpETAF78Wpas:-KNYnwqDYq_HLVmUytCY:14315048) ([raw file](https://github.com/cockroachdb/cockroach/blob/2dde8c3c84ad25e7018f8a0762825b7c6b781512/security/tls.go#L122)):*
<details><summary><i>Previously, d4l3k (Tristan Rice) wrote…</i></summary>
> Sorry for the poor wording. `(*tls.Config).Clone()`is also slated for 1.8. When that gets added, we'll remove this function.
</details>
add a link to the upstream implementation.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7949#-:-KNYm0us40Yb5EAEgXiE:1897477994)*

**Comment by tschottdorf** [comment]
Reviewed 5 of 5 files at r1, 9 of 9 files at r2, 10 of 10 files at r3, 9 of 9 files at r4, 1 of 1 files at r5, 2 of 2 files at r6, 30 of 30 files at r7, 9 of 9 files at r8, 1 of 1 files at r9, 2 of 2 files at r10.
Review status: all files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7949)*

**Pull request closed without merging**